### ISSUE TITLE
Precision for matmul apply()

### ISSUE THREAD

**Issue by cho-uc** [ginkgo-project/ginkgo#741]
There could be a bug in the `apply()` and `apply2()` method if it's run with elements with fine precision.
I had a simple test program for matrix multiplication using 1000 elements run in Tesla V100 using double precision.
I compared the result (using summation of all elements) with CPU calculation that I wrote myself using `std::vector`.

Both of them should give the same result if they're run with the same double precision, but they didn't.
I discovered this when I wrote another application that's quite sensitive with this rounding error. This program will not work if  it's not precise

What could be the explanation?

Here is the link to the code :
https://github.com/cho-uc/ginkgo_test

**Comment by upsj** [comment]
Thanks for your report and detailed reproduction example! You are raising a good point, that is even when given the same input data, different matrix formats may give slightly different (and even non-deterministic) results in their SpMV calculations. We should probably note that more clearly in our documentation.

### Floating point associativity

I'm don't know how familiar you are with floating point arithmetic, **so you may skip this part** if you already know about the issues with associativity:

If *u* is the smallest number that you can add to 1 to still get a number larger than 1 (for double precision, u is around 2*10^-16), then

> (1 + u/2) + u/2 = 1 + u/2 = 1,

while

> 1 + (u/2 + u/2) = 1 + u > 1,

so depending on the order in which we accumulate partial sums in a floating point number type matters, which is why floating point addition is not associative. Now when computing a matrix-vector product, we are summing up
> a_11 x_1 + a_12 x_2 + a_13 x_3 + a_14 x_4

and as above, changing the order of evaluation (i.e. the parentheses I left out here) can give you slightly different results:

> (((a_11 x_1 + a_12 x_2) + a_13 x_3) + a_14 x_4) =/= (a_11 x_1 + (a_12 x_2 + (a_13 x_3 + a_14 x_4)))

### How Ginkgo SpMVs (matmul) accumulate their sums

Our GPU implementations use different kinds of accumulation algorithms: Some of them are equivalent to sequential accumulation like some (but not all) instances of ELL matrices (`num_thread_per_worker == 1`):

> (((a_11 x_1 + a_12 x_2) + a_13 x_3) + a_14 x_4)

some of them use a parallel accumulation following a balanced binary tree, like some CSR configurations (classical strategy), COO and ELL (`num_thread_per_worker > 1`) implementations to accumulate partial results
> ((a_11 x_1 + a_12 x_2) + (a_13 x_3 + a_14 x_4))

or use atomic operations to get something completely non-deterministic, like some CSR, ELL and COO implementations for adding up the partial results to the total sum
> ((a_11 x_1 + (a_12 x_2 + a_13 x_3)) + a_14 x_4)

**The bottom line is:** all of our matrix formats may give slightly different or even non-deterministic results when used on GPUs. It may well be that there is a bug within our COO implementation, but this would be the most obvious explanation I can think of. I will still investigate your example a bit further to see if anything is really going wrong in our code.

**Comment by upsj** [comment]
For your application: Can you give a bit more detail on the computations you are doing and where this sensitivity to rounding errors comes from? Maybe we can figure out another way to make your computations more stable.
Until then, the OpenMP implementations of all formats except COO should give you adequatly good performance with bitwise equivalence to the sequential implementation.

**Comment by cho-uc** [comment]
The computations that I was doing was solving  b= A.x for _Finite Cell Method_ using _Geometric Multigrid_ in GPU.
I used _Additive-Schwarz_ as preconditioner. It was noted in some research paper that this preconditioner is unstable so it could be that the small rounding error is affecting the output of the program greatly.
Unfortunately I can't switch OpenMP because of the requirement of the project.

Thanks for trying to investigating this!

**Comment by upsj** [comment]
I'll just throw out a few questions, maybe they help pinpoint what might be the issue :) What kind of PDE are you solving? Regarding Multigrid, are you using an F-cycle? Or V-cycle with a suitable initial guess? What kind of coarsest grid solver and smoother are you using? Are you able to solve the problem correctly with a hand-written CPU implementation?

**Comment by cho-uc** [comment]
* It's a simple Poisson equation as a benchmark problem. We already had parallel version running on CPU (with Petsc). My responsibility is to port it to GPU. Anyway, I am still not sure what's the reason that this program didn't work. It worked on some benchmark problem, but for other configuration (for example for finer mesh) didn't work if we changed the configuration.
That's why I thought that it could be due to the rounding error. But there could be other causes as well.
* it's a V cycle
* Jacobi smoother
* I use GMRES solver with ILU preconditioner
* The problem can be solved correctly with the CPU implementation

**Comment by upsj** [comment]
That does indeed look like a hard problem to pinpoint. It also sounds like you would not be able to share/give access to your full prototype code directly? Before going deeper into this issue, you are "solving" an error equation (Ae = b) on each level and applying the error on the solution afterwards (x = x + e), correct? Do you know how the norm of the residual of this equation (b - Ae) looks initially, after pre-smoothing, the actual V-cycle and after post-smoothing in the CPU and GPU case? What convergence level do you reach in the GMRES solver?
Note that the default ILU preconditioner uses only an approximation to the exact ILU(0), so it might help to use an exact ILU factorization for the preconditioner (`.with_factorization_factory(gko::preconditioner::Ilu::build().on(exec))`), which we should maybe also make the default behavior everywhere as soon as we have matching CPU implementations for it.

**Comment by cho-uc** [comment]
Roughly you're correct.
See
https://github.com/cho-uc/ginkgo_test/blob/main/GMG_algo.png
for the GMG algorithm.
Unfortunately I can't share the prototype.
Related to your questions about the base solver, here is the code that we use
https://github.com/cho-uc/ginkgo_test/blob/main/ilu-preconditioned-solver.cpp

Does it belong to an exact ILU factorization or just approximation?

Anyway, we have solved the issue in our program, and the error is not because of the rounding error from GPU.
Moving forward, it will be nice if this rounding error in Ginkgo is documented properly, as previously we assume that all the computation from GPU is the same with CPU version down to the last decimal digits.

**Comment by upsj** [comment]
Thanks, that clears up things! We will try to clarify this part of our documentation.

The code you posted uses the approximate ILU factorization. If you want to switch to the exact one, replace `ParIlu` by `Ilu` in https://github.com/cho-uc/ginkgo_test/blob/27e5131ab14be7e09725396dc4d1f1881a60d7aa/ilu-preconditioned-solver.cpp#L28
(note that right now, this only works on the GPU, we haven't implemented CPU support yet)

**Comment by upsj** [comment]
I'll close this issue to create a more appropriate one for the necessary changes.
